[DataGridPro] Add data source interface and basic documentation (#10543)
mui/mui-x: docs/data/data-grid/server-side-data/index.md

The data grid is configured to fetch data from the server whenever the user changes the page or updates filtering or sorting.

```tsx
const [rows, setRows] = React.useState([]);
const [paginationModel, setPaginationModel] = React.useState({
  page: 0,
  pageSize: 10,
});
const [filterModel, setFilterModel] = React.useState({
  items: [],
});
const [sortModel, setSortModel] = React.useState([]);

React.useEffect(() => {
  const fetcher = async () => {
    // fetch data from server
    const data = await fetch('https://my-api.com/data', {
      method: 'GET',
      body: JSON.stringify({
        page: paginationModel.page,
        pageSize: paginationModel.pageSize,
        sortModel,
        filterModel,
      }),
    });
    setRows(data.rows);
  };
  fetcher();
}, [page, sortModel, filterModel]);

<DataGridPro
  columns={columns}
  pagination
  sortingMode="server"
  filterMode="server"
  paginationMode="server"
  onPaginationModelChange={(newPaginationModel) => {
    setPaginationModel(newPaginationModel);
  }}
  onSortModelChange={(newSortModel) => {
    setSortModel(newSortModel);
  }}
  onFilterModelChange={(newFilterModel) => {
    setFilterModel(newFilterModel);

>   }}

Reviewer: These could be written without the added wrapper, gives a more correct example.

The datasource will work with all the major data grid features which require server-side data fetching such as sorting, filtering, pagination, grouping, etc.

### Usage

The data grid server-side data source has an initial set of required methods that you need to implement. The data grid will call these methods internally when the data is required for a specific page.

```tsx
interface DataSource {
  /**
    Fetcher Functions:
    - `getRows` is required
    - `updateRow` is optional
    
    `getRows` will be used by the grid to fetch data for the current page or children for the current parent group
    It may return a `rowCount` to update the total count of rows in the grid
  */
  getRows(params: GetRowsParams): Promise<GetRowsResponse>;
  // if passed, will be called like `processRowUpdate` on row edit
  updateRow?(updatedRow: GridRowModel): Promise<any>;
}
```

Here's how the code will look like for the above example:

```tsx

> const myCustomDataSource = {

Reviewer: lol I would actually not use the explicit typing, the name is clear enough. Pet peeve but I'd also remove the `my`, sounds quite dated/perly to my ear.

^ All nitpicks, feel free to ignore

  will be used by the grid to group rows by their parent group
  This effectively replaces `getTreeDataPath`.
  Consider this structure:

  ```
  - (1) Sarah // groupKey 'Sarah'
    - (2) Thomas // groupKey 'Thomas'
  ```

  When (2) is expanded, the `getRows` function will be called with group keys `['Sarah', 'Thomas']`.

- `hasChildren?(row: GridRowModel): boolean`

  Will be used by the grid to determine if a row has children on server

- `getChildrenCount?: (row: GridRowModel) => number`

  Will be used by the grid to determine the number of children of a row on server

#### Existing server-side features

The server-side `dataSource` will change a bit the way existing server-side features work. Whenever a valid data source is passed the features `filtering`, `sorting`, `pagination` will automatically be set to `server`.

However, if for some reason, you want to explicitly work one or more of these features to work on `client`, you can do so by passing the `client` value to the feature. Here's an example:

```tsx
<DataGridPro
  columns={columns}
  dataSource={myCustomDataSource} // will set `filtering`, `sorting`, `pagination` to `server`

>   filterMode="client" // will set `filtering` to `client` i.e. data won't be refetched on filter change

Author: I guess yes, it will work like how the client-side filter works today, i.e when the rows prop update, or `updateRows` is called, the filter is persisted and applied on the new rows, likewise if the data is refetched or new rows are fetched they'll be subjected to the active client-side filtering rules.

Although this won't be a need for many users, I'm sure there will be use cases where people will want this.

>   filterMode="client" // will set `filtering` to `client` i.e. data won't be refetched on filter change

Reviewer: I'm curious, what would be the use-case for this? I have trouble making sense of server data with client filtering.

>   filterMode="client" // will set `filtering` to `client` i.e. data won't be refetched on filter change

Author: > Is this assuming that all the server rows are fetched?

No, actually it's adressing those rare use-cases when let's say filtering has yet not implemented on the server and the developer wants to keep it working on the client-side on the partially server loaded data. Tbh, I'm also not 100% sure about this. It has been asked as a [question](https://www.notion.so/Data-source-draft-db58b8b749154420a407f231c5528c1d?d=407dd9c21ac349e59fc0926afe0e4443&pvs=4#9c8f2af9613b48089c89fd0b12c4d17a) in the data source proposal, which made me think it might be a use-case requested by some users.

>   filterMode="client" // will set `filtering` to `client` i.e. data won't be refetched on filter change

Reviewer: I'm not sure if that use-case makes sense. Wouldn't it be confusing for end-users, that the fetching is server-side but the filtering is client-side? And couldn't it create weird UX events, e.g. if there is filtering + lazy-loading, and the lazy-loaded rows are filtered out as it loads?

I feel like it might be desirable to avoid/restrict this use-case.

>   filterMode="client" // will set `filtering` to `client` i.e. data won't be refetched on filter change

Author: Yeah, good point around the possibly weird UX experience.
Alright, since we are anyways not sure about the need for this, let's not do it until a demand is there from some users for this specifically. 👍

/>
```

#### Caching

The data grid will cache the data it receives from the server. This means that if the user navigates to a page that has already been fetched, the grid will not call the `getRows` function again. This is to avoid unnecessary calls to the server.

> By default, the data grid will use the tanstack query library to cache the data. A `QueryClient` will be created internally and used with the data source. You can also pass your own instance of `QueryClient` with a optional custom configuration and use it to play with the cached data. You can also use this client to invalidate the cache when you need to.

Reviewer 2: ```suggestion
By default, the data grid will use the `@tanstack/react-query` library to cache the data. A `QueryClient` will be created internally and used with the data source. You can also pass your own instance of `QueryClient` with a optional custom configuration and use it to play with the cached data. You can also use this client to invalidate the cache when you need to.
```

Reviewer: For my information, why did we decide to use react-query?

Author: To take advantage of the powerful caching functionality it provides and allow users the customization with caching that it provides.

Author: We could also think of making this library agnostic in future so that users are free to use:
- No cache
- custom made cache
- react-query
- swr
- any other that they want

But I feel that should happen if there's a strong need for this kind of customization and people are demanding it. As far as I think, most of the caching use cases could be covered with tanstack query.

Reviewer 2: > we expose an external library as an API so we're a bit stuck with it

That's a good point. Also, we'll only use a part of the library and I'm not sure how well will tree shaking work in this case.

> For something that in the end doesn't seem terribly more than just a Map<requestId, Response>.

Yes, plus cache configuration and proper invalidation.

I considered using react query as a shortcut, but now I feel more convinced to our own implementation 🙂

Reviewer: I'm still not convinced.

I think we need that for CSS because users come in with their styling solution already. It doesn't apply for fetching clients. And if we expose a library-agnostic interface, then it makes absolutely no difference for users what the underlying client is.

But I'll note that if we could expose an opaque or library-agnostic interface, then I wouldn't really mind using react-query internally for the next release - because we would still have the liberty of improving it anytime.

What are the use-cases we're trying to solve by exposing that object? Do you have any pseudo-code or doc examples I could look at to understand better?

Author: > What are the use-cases we're trying to solve by exposing that object? Do you have any pseudo-code or doc examples I could look at to understand better?

I'm sure there could be multiple ones, but one of the prominent use-cases we were targeting is the easy synchronization with the other apps like toolpad or charts.
An easy way out that cache from the library can provide is being able to subscribe to the changes and get the updated values from the cache like shown in this example: https://codesandbox.io/s/datagridpro-basic-forked-49zpsy?file=/src/App.tsx (but yes, that could be done with the custom implementation)

> But I'll note that if we could expose an opaque or library-agnostic interface, then I wouldn't really mind using react-query internally for the next release - because we would still have the liberty of improving it anytime.

Thanks for sharing the useful thoughts on this. I didn't think like this previously, I seem to agree with you that we shouldn't expose the interface of an external library we don't control. And I believe it might get a bit trickier to nail, but we can solve the problems like cache data sharing and request deduplication internally.
I'll update the documentation and see how it looks like for an end user.

As for the customization the library provided, in the beginning, we can even chose to not allowing any customization on the user's side, to follow it up later as per the needs of the users.

Author: I've removed the implementation details and interface around cache for now, currently decision is to not directly expose a library and try to do internal implementation for caching. Resolving the conversation, feel free to unresolved if some point still needs discussion.

  onFilterModelChange={setFilterModel}
/>;
```

We only scratched the surface in the above example with a lot of problems still unsolved like:

- Performance optimization
- Caching data/deduping requests
- More complex use-cases on the server like grouping, tree data, etc.
- Server side row editing
- Lazy loading of data
- Handling updates to the data like row editing, row deletion, etc.
- Refetching data on-demand

Trying to solve these problems one after the other can make the code complex and hard to maintain.

## Data source

A very common pattern to solve these problems is to use a centralized data source. A data source is an abstraction layer that sits between the data grid and the server. It provides a simple interface to the data grid to fetch data and update it. It handles a lot of the complexities related to server side data fetching. We will talk more about the data source in the next section.

:::warning

This feature is still <b>under development</b> and the information shared on this page is subject to change. Feel free to subscribe or comment on the official GitHub [issue](https://github.com/mui/mui-x/issues/8179).

:::

### Overview

> The Data Grid already supports manual server-side data fetching for features like sorting, filtering, etc. In order to make it more powerful and simple to use, we provide a data source interface that you can implement with your existing data fetching logic.

Author: ```suggestion
The Data Grid already supports manual server-side data fetching for features like sorting, filtering, etc. In order to make it more powerful and simple to use, the grid will support a data source interface that you can implement with your existing data fetching logic.
```

The datasource will work with all the major data grid features which require server-side data fetching such as sorting, filtering, pagination, grouping, etc.

### Usage

The data grid server-side data source has an initial set of required methods that you need to implement. The data grid will call these methods internally when the data is required for a specific page.

```tsx

> interface DataSource {

Author: Yep, nice catch, thanks!

   * `hasNextPage`: When row count is unknown/inaccurate, if `truncated` is set or rowCount is not known, data will keep loading until `hasNextPage` is `false`
   * `truncated`: To reflect `rowCount` is inaccurate (will trigger `x-y of many` in pagination after the count of rows fetched is greater than provided `rowCount`)
   * It could be useful with:
   * 1. Cursor based pagination:
   *   When rowCount is not known, grid will check for `hasNextPage` to determine
   *   if there are more rows to fetch.
   * 2. Inaccurate `rowCount`:
   *   `truncated: true` will let the grid know that `rowCount` is estimated/truncated.
   *   Thus `hasNextPage` will come into play to check more rows are available to fetch after the number becomes >= provided `rowCount`
   */
  pageInfo?: {
    hasNextPage?: boolean;
    truncated?: number;
  };
}
```

#### Updating data

If provided, the method `dataSource.updateRow` will be called with the `GridRowModel` object whenever the user edits a row. This method is optional and you can skip it if you don't need to update the data on the server. It will work in a similar way as the `processRowUpdate` prop.

> #### Data grid props

Author: ```suggestion
#### Data Grid props
```

These data grid props will work with the server-side data source:

- `dataSource: DataSource`: the data source object that you need to implement
- `rows`: will be ignored, could be skipped when `dataSource` is provided

> - `rowCount`: will be used to identify the total number of rows in the grid, if not provided, the grid will use the _GetRowsResponse.rowCount_ value

Author: ```suggestion
- `rowCount`: will be used to identify the total number of rows in the grid, if not provided, the grid will check for the _GetRowsResponse.rowCount_ value, unless the feature being used is infinite loading where no `rowCount` is available at all.
```
